src/models/Throttler.ts

```diff
@@ -1,45 +1,209 @@
+import { Message } from "discord.js";
 import { getTimeLeft } from "../utils/timeout";
 
-export class Throttler {
-    private _current = 0;
-    private _timeout: NodeJS.Timeout | undefined = undefined;
+export type ThrottlerScope = "global" | "user" | "member" | "guild";
 
+export interface CommandThrottler {
+    /** The scope of the throttler. */
+    readonly scope: ThrottlerScope;
+    /** How many time the throttler can be triggered. */
+    readonly count: number;
+    /** Time in seconds since the first trigger before the throttler is reset. */
+    readonly duration: number;
+    /**
+     * Gets the number of time the throttler has been triggered.
+     * @param message
+     */
+    getCurrent(message: Message): number;
+    /**
+     * Gets the number of time the throttler has been triggered.
+     * @param id
+     */
+    getCurrent(id: string): number;
+    /**
+     * Gets either or not the throttler has reached the limit.
+     * @param message
+     */
+    getThrottled(message: Message): boolean;
+    /**
+     * Gets either or not the throttler has reached the limit.
+     * @param id
+     */
+    getThrottled(id: string): boolean;
+    /**
+     * Gets the time in seconds until the throttler is reset.
+     * @param message
+     */
+    getCooldown(message: Message): number;
+    /**
+     * Gets the time in seconds until the throttler is reset.
+     * @param id
+     */
+    getCooldown(id: string): number;
+    /**
+     * Resets all throttlers.
+     */
+    reset(): void;
     /**
+     * Resets the throttler.
+     * @param message
+     */
+    reset(message: Message): void;
+    /**
+     * Resets the throttler.
+     * @param id
+     */
+    reset(id: string): void;
+    /**
+     * Increment the throttler's counter.
+     * @param message
+     * @returns Either or not the limit has been reached.
+     * @example
+     * ```ts
+     * if (throttler.add(message)) {
+     *     // Do action
+     * } else {
+     *     // Cannot do action
+     * }
+     * ```
+     */
+    add(message: Message): boolean;
+    /**
+     * Increment the throttler's counter.
+     * @param id
+     * @returns Either or not the limit has been reached.
+     * @example
+     * ```ts
+     * if (throttler.add(id)) {
+     *     // Do action
+     * } else {
+     *     // Cannot do action
+     * }
+     * ```
+     */
+    add(id: string): boolean;
```

Comment:
```suggestion
    /**
     * Increment the throttler's counter.
     * @param message
     * @returns Either or not the limit has been reached.
     * @example
     * ```ts
     * if (throttler.increment(message)) {
     *     // Do action
     * } else {
     *     // Cannot do action
     * }
     * ```
     */
    increment(message: Message): boolean;
    /**
     * Increment the throttler's counter.
     * @param id
     * @returns Either or not the limit has been reached.
     * @example
     * ```ts
     * if (throttler.increment(id)) {
     *     // Do action
     * } else {
     *     // Cannot do action
     * }
     * ```
     */
    increment(id: string): boolean;
```